### PR TITLE
qdel 1234 1234 -> server core dump

### PR DESCRIPTION
**Pull request by ericpershey**
#### Describe Bug or Feature
When you qdel the same jobid more than once in a single qdel, it can segfault the server.

#### Describe Your Change
* Made changes in an attempt to mitigate the problem; however the changes do not handle batches of job/subjob deletes.  The solution for handling batches is unclear.
* Added a series of tests to expose the problem.

#### stack trace of the segfault
    pbsdev@pdw-s1:~ 21% sudo gdb /opt/pbs/sbin/pbs_server.bin /var/spool/pbs/server_priv/core_0001
    (gdb) bt
    #0  0x000014900fc2aa9f in raise () from /lib64/libc.so.6
    #1  0x000014900fbfde05 in abort () from /lib64/libc.so.6
    #2  0x000014900fc6d037 in __libc_message () from /lib64/libc.so.6
    #3  0x000014900fc7419c in malloc_printerr () from /lib64/libc.so.6
    #4  0x000014900fc75f0d in _int_free () from /lib64/libc.so.6
    #5  0x00000000004cd80c in free_string_array (arr=0x16f4600) at /home/pbsdev/pbs/source/src/lib/Libutil/misc_utils.c:1346
    #6  0x00000000004537c8 in free_br (preq=0x166f700) at /home/pbsdev/pbs/source/src/server/process_request.c:1633
    #7  0x00000000004558f4 in reply_send (request=0x166f700) at /home/pbsdev/pbs/source/src/server/reply_send.c:369
    #8  0x0000000000455d95 in req_reject (code=15001, aux=0, preq=0x166f700) at /home/pbsdev/pbs/source/src/server/reply_send.c:536
    #9  0x000000000045703e in req_deletejob (preq=0x166f700) at /home/pbsdev/pbs/source/src/server/req_delete.c:624
    #10 0x00000000004525b3 in dispatch_request (sfds=19, request=0x166f700) at /home/pbsdev/pbs/source/src/server/process_request.c:856
    #11 0x00000000004522af in process_request (sfds=19) at /home/pbsdev/pbs/source/src/server/process_request.c:695
    #12 0x00000000004e00d1 in process_socket (sock=19) at /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:510
    #13 0x00000000004e03e8 in wait_request (waittime=2, priority_context=0x165ceb0) at /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:623
    #14 0x0000000000450a74 in main (argc=1, argv=0x7ffdb86efc88) at /home/pbsdev/pbs/source/src/server/pbsd_main.c:1396

#### ASAN of specific code.
    ==pbs_server.bin==57676==ERROR: AddressSanitizer: heap-use-after-free on address 0x61d0000564c8 at pc 0x000000450694 bp 0x7ffd4f21fb00 sp 0x7ffd4f21f288
    READ of size 7 at 0x61d0000564c8 thread T0
        #0 0x450693 in printf_common(void*, char const*, __va_list_tag*) /home/abuild/rpmbuild/BUILD/llvm-11.0.1.src/build/../projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors_format.inc:547:9
        #1 0x450cb7 in vsprintf /home/abuild/rpmbuild/BUILD/llvm-11.0.1.src/build/../projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:1668:1
        #2 0x451d63 in sprintf /home/abuild/rpmbuild/BUILD/llvm-11.0.1.src/build/../projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:1711:1
        #3 0x664209 in acct_del_write /home/pbsdev/pbs/source/src/server/req_delete.c:198:2
        #4 0x66a70f in post_delete_mom1 /home/pbsdev/pbs/source/src/server/req_delete.c:1662:2
        #5 0x956413 in dispatch_task /home/pbsdev/pbs/source/src/lib/Libutil/work_task.c:184:3
        #6 0x57acd5 in process_DreplyTPP /home/pbsdev/pbs/source/src/server/issue_request.c:873:5
        #7 0x5d2753 in is_request /home/pbsdev/pbs/source/src/server/node_manager.c:4309:4
        #8 0x62d6f8 in do_tpp /home/pbsdev/pbs/source/src/server/pbsd_main.c:310:4
        #9 0x62d8ca in tpp_request /home/pbsdev/pbs/source/src/server/pbsd_main.c:353:3
        #10 0x9839d3 in process_socket /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:510:2
        #11 0x9827c9 in wait_request /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:623:8
        #12 0x635888 in main /home/pbsdev/pbs/source/src/server/pbsd_main.c:1396:7
        #13 0x145eef3522bc in __libc_start_main (/lib64/libc.so.6+0x352bc)
        #14 0x432259 in _start /home/abuild/rpmbuild/BUILD/glibc-2.31/csu/../sysdeps/x86_64/start.S:120

    0x61d0000564c8 is located 72 bytes inside of 2280-byte region [0x61d000056480,0x61d000056d68)
    freed by thread T0 here:
        #0 0x4ac14d in free /home/abuild/rpmbuild/BUILD/llvm-11.0.1.src/build/../projects/compiler-rt/lib/asan/asan_malloc_linux.cpp:123:3
        #1 0x63e90e in free_br /home/pbsdev/pbs/source/src/server/process_request.c:1687:9
        #2 0x653c5a in reply_send /home/pbsdev/pbs/source/src/server/reply_send.c:369:2
        #3 0x654393 in reply_ack /home/pbsdev/pbs/source/src/server/reply_send.c:412:9
        #4 0x66a85b in post_delete_mom1 /home/pbsdev/pbs/source/src/server/req_delete.c:1668:4
        #5 0x956413 in dispatch_task /home/pbsdev/pbs/source/src/lib/Libutil/work_task.c:184:3
        #6 0x57acd5 in process_DreplyTPP /home/pbsdev/pbs/source/src/server/issue_request.c:873:5
        #7 0x5d2753 in is_request /home/pbsdev/pbs/source/src/server/node_manager.c:4309:4
        #8 0x62d6f8 in do_tpp /home/pbsdev/pbs/source/src/server/pbsd_main.c:310:4
        #9 0x62d8ca in tpp_request /home/pbsdev/pbs/source/src/server/pbsd_main.c:353:3
        #10 0x9839d3 in process_socket /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:510:2
        #11 0x9827c9 in wait_request /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:623:8
        #12 0x635888 in main /home/pbsdev/pbs/source/src/server/pbsd_main.c:1396:7
        #13 0x145eef3522bc in __libc_start_main (/lib64/libc.so.6+0x352bc)

    previously allocated by thread T0 here:
        #0 0x4ac3cd in malloc /home/abuild/rpmbuild/BUILD/llvm-11.0.1.src/build/../projects/compiler-rt/lib/asan/asan_malloc_linux.cpp:145:3
        #1 0x63ac37 in alloc_br /home/pbsdev/pbs/source/src/server/process_request.c:1245:33
        #2 0x637ff8 in process_request /home/pbsdev/pbs/source/src/server/process_request.c:412:17
        #3 0x9839d3 in process_socket /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:510:2
        #4 0x9827c9 in wait_request /home/pbsdev/pbs/source/src/lib/Libnet/net_server.c:623:8
        #5 0x635888 in main /home/pbsdev/pbs/source/src/server/pbsd_main.c:1396:7
        #6 0x145eef3522bc in __libc_start_main (/lib64/libc.so.6+0x352bc)

### CODE REVIEW

File: test/tests/functional/pbs_qdel.py [hunk 1]
```diff
@@ -164,3 +167,292 @@ def test_qdel_hstry_jobs_rerun(self):
...
+        # once.
+        try:
+            self.server.deljob([jid, jid, jid, jid])
+        except PbsDeljobError as e:
+            err_msg = "could not connect to MOM"
+            self.assertTrue(err_msg in e.msg[0],
+                            "Did not get the expected message")
+            self.assertTrue(e.rc != 0, "Exit code shows success")
+        else:
+            raise self.failureException("qdel job did not return error")
```
Comment (bhroam) [inline test/tests/functional/pbs_qdel.py:197]: unittest has an assert for this exact purpose.
Try:
```
with self.assertRaises(PbsDeljobError, msg="qdel job did not return error") as e:
    self.server.deljob([jid, jid, jid, jid])

self.assertIn(err_msg, e.exception.msg[0])
self.assertNotEqual(e.exception.rc, 0)
```

File: src/server/req_delete.c [hunk 2]
```diff
@@ -866,6 +878,10 @@ req_deletejob(struct batch_request *preq)
 			chk_array_doneness(parent);
 		}
 	}
+
+fn_exit:
```
Comment (bhroam) [inline src/server/req_delete.c:882]: Two things about this label.  First, labels are scoped to functions.  This means you don't need to prefix it if you don't want to.  Second, I wouldn't call it exit.  You aren't exiting the program, you're basically returning.

I've seen 'done' used as a label name in other parts of the code.

In any case, the label name is fine if you want to keep it.

Comment (toonen) [reply src/server/req_delete.c:882]: I generally prefix the label with the scope to make it more, hence the 'fn' for function.  Your point about 'exit' is valid and I normally use fn_return and fn_error, the latter for handling exception like situations.  I'll rename the label once we figure out the correct solution.  The changes to req_delete.c are only a partial solution.  The main purpose of the PR was to submit failing tests and initiate a discussion on the right way to solve the issues.

**Comment by toonen** [comment]
Hi @nithinj.

Thanks for the description of the data structures involved and why the crash occurs.  What you said matches @ericpershey and my understanding of the issue.

We tried using the PBS_IDX_DUPS_OK flag early on in our attempts to fix the issue; however the implementation of PBS_IDX_DUPS_OK, or more specifically AVL_DUP_KEYS_OK, appears to be broken or perhaps just unfinished.   We looked for a more up-to-date version of avltree.[ch] on the web but were unable to find one.

Clearing duplicate normal and array job ids from the request list would be simple enough using the AVL tree to detect the duplicates.  Unfortunately, that solution is insufficient for subjobs, and especially overlapping subjob ranges.  Perhaps something like an interval tree could be used to replace the subjob specifications with ranges that are nonoverlapping.

Our current solution does not solve all of the issues identified by the tests we added.  In addition, the current solution does not solve batching of deletes when the number of jobs/subjobs is sufficient large to warrant batching.  One of our attempted solutions was to bump the reference count on the batch request for outstanding MoM operations, but we didn't succeed in making that work; athough now that we understand the code a little better, perhaps another reference counting attempt is warranted.

As for the sanity checking in acct_del_write(), did you have anything specific in mind?

**Comment by bhroam** [comment]
I'm not sure it will help, but we have code to handle subjob ranges.  Maybe something can be made using that?  It's in miscutils called range.c.  It doesn't have anything to do with jobs or batch structures.  It's a new structure of its own that handles ranges.  I originally wrote it for the scheduler to handle subjobs, but at some point we made it more generic and moved it into misc utils.

**Comment by toonen** [comment]
ALL: I'm going to change this PR to a work in progress (WIP) so that it doesn't accidentally get merged.  As said previously, the server code changes provided don't solve the issues in entirety and may not even be on the correct path to a solution.  The primary point of this PR was to identify the issue, provide a set of tests that need to be solved or prove to be bad tests, and most importantly start a discussion about possible solutions.

File: test/tests/functional/pbs_qdel.py (hunk 1 above)
Comment (ericpershey) [reply test/tests/functional/pbs_qdel.py]: Done

**Review by nithinj — APPROVED** [review]
LGTM!

**Review by subhasisb — APPROVED** [review]

**Comment by subhasisb** [comment]
please resolve the conflicts and push again

**Comment by toonen** [comment]
@subhasisb I merged in the changes from master into this PR.  Since PR #2581 eliminated the duplicate job ids, I removed the earlier changes to the server but left the tests.  The tests exposed a couple of issues with the PR #2581 changes.  I believe I fixed those in commit e82c0af.  With the exception of test_qdel_history_job, all of the tests in TestQdel now cleanly pass when built with Address Sanitizer.

test_qdel_history_job fails when PBS is built with ASan, but that is also true for the master branch and has been for quite some time.   I can file an issue with more details if it would help, but it appears to be a result of some pointer math that never made sense to me for resources that are not built-ins.  The following appears to be the relevant output from ASan.

`
src/lib/Libattr/attr_fn_resc.c:506:10: runtime error: subtraction of unsigned offset from 0x000000cedfe0 overflowed to 0x6070001767b0
`

**Comment by riyazhakki** [comment]
@toonen, Thank you for adding the testcases and updating the code.
Code changes looks good to me.

**Review by riyazhakki — APPROVED** [review]

**Comment by riyazhakki** [comment]
> test_qdel_history_job fails when PBS is built with ASan, but that is also true for the master branch and has been for quite some time. I can file an issue with more details if it would help, but it appears to be a result of some pointer math that never made sense to me for resources that are not built-ins. The following appears to be the relevant output from ASan.
> 
> `src/lib/Libattr/attr_fn_resc.c:506:10: runtime error: subtraction of unsigned offset from 0x000000cedfe0 overflowed to 0x6070001767b0`


The problem seems unrelated, but would you be able to provide valgrind logs?

**Comment by toonen** [comment]
@riyazhakki yeah.  the test_qdel_history_job failure is unrelated, and only occurs if PBS is built with Address Sanitizer.  I just wanted you all to be aware of it.  It should not be a reason to stop the merge.

**Pull request merged**